src/Microsoft.Data.Analysis.Interactive/DataFrameKernelExtension.cs

```diff
@@ -1,47 +1,140 @@
+            if (type.IsArray)
+            {
+                // Not handled yet
+                return "DataFrameColumn";
+            }
+            if (type.IsGenericType)
+            {
+                int backTick = friendlyName.IndexOf('`');
+                if (backTick > 0)
+                {
+                    friendlyName = friendlyName.Remove(backTick);
+                }
+                friendlyName += "<";
+                Type[] typeParameters = type.GetGenericArguments();
+                for (int i = 0; i < typeParameters.Length; ++i)
+                {
+                    string typeParamName = GetFriendlyName(typeParameters[i]);
+                    friendlyName += (i == 0 ? typeParamName : "," + typeParamName);
+                }
+                friendlyName += ">";
+            }
+
+            return friendlyName;
+        }
+
+        public string GetTypedDataFrameWithProperties(DataFrame dataFrame, string resultTypeName, out StringBuilder prettyFormatter)
+        {
+            prettyFormatter = new StringBuilder();
+            StringBuilder stringBuilder = new StringBuilder();
+            string constructor = @$"
+public class {resultTypeName} : DataFrame 
+{{
+    public {resultTypeName}(DataFrame dataFrame)
+    {{
+        foreach (var column in dataFrame.Columns)
+        {{
+            Columns.Add(column);
+        }}
+    }}
+
+";
+            stringBuilder.Append(constructor);
+            prettyFormatter.Append(constructor);
 
+            foreach (var column in dataFrame.Columns)
+            {
+                string columnName = column.Name.Replace(" ", string.Empty);
+                Type dataType = column.DataType;
+                string typeName = GetFriendlyName(column.GetType());
+                stringBuilder.Append($@"
+    public {typeName} {columnName} 
+    {{
+        get
+        {{
+            int columnIndex = Columns.IndexOf(""{columnName}"");
+            return Columns[columnIndex] as {typeName};
+        }}
+    }}");
+                stringBuilder.AppendLine();
+                prettyFormatter.Append($@"
+    public {typeName} {columnName} {{ get; }}");
+                prettyFormatter.AppendLine();
+            }
+
+            stringBuilder.AppendLine();
+            stringBuilder.Append(@"}");
+            prettyFormatter.AppendLine();
+            prettyFormatter.Append(@"}");
+            prettyFormatter.AppendLine();
+            return stringBuilder.ToString();
+        }
+
+        public Task OnLoadAsync(IKernel kernel)
+        {
             var kernelBase = kernel as KernelBase;
-            var directive = new Command("#!doit")
+            var directive = new Command("#!AddStrongTypes")
             {
-                Handler = CommandHandler.Create(async (FileInfo csv, string typeName, KernelInvocationContext context) =>
+                Handler = CommandHandler.Create(async (string dataFrameName, string typeName, KernelInvocationContext context) =>
```

Comment:
I'm curious, why did you remove the csv argument from the command arguments? The current approach requires 3 commands to initialize a frame with strongly typed columns, shouldn't it be possible to do it with just one?

Reply:
Agreed I think we took a step back in usability here. I think it is a good addition to add the option to take an existing `DataFrame` as input, as that could be useful. But I would expect a "simple" case to still be possible.

My thinking on the simple case would be:

```
#!LoadDataFrame csv C:\blah\housing.csv
```
```
housing.Description()
```

Note, we can try to infer both the generated type and the variable name from the file name. In this case, we would generate `HousingDataFrame` Type and `housing` variable which is populated with the data from that file.

Reply:
Hmmm, my initial thought was for the magic command to generate strongly typed columns on any DataFrame. So in my head, I kinda separated out reading a csv file from generating a new type. I'll add the csv option back in.

Reply:
Alright, so our `LoadCsv` method can sometimes fail to infer the correct types. In these cases, this extension will generate the wrong column types for it's properties, but this is ok since we can pass in a regular `DataFrame` and generate strong properties on it.